### PR TITLE
LaTeX compiler in-browser

### PR DESCRIPTION
**Pull request by pulsejet**
Supports compilation of LaTeX in browser with WASM (https://github.com/SwiftLaTeX/SwiftLaTeX)

This is mostly just a PoC not intended for merging due to limitations in SwiftLaTeX:
1. It's way of fetching required files is terribly inefficient (no caching)
2. It also relies on a service hosted by SwiftLaTeX to fetch these files instead of CTAN (even though I read somewhere that it _can_ use CTAN ...)
3. SwiftLaTeX is AGPL-3.0

image

### CODE REVIEW

**Comment by zjkmxy** [comment]
#5

**Review by zjkmxy — APPROVED** [review]
LGTM

**Comment by yoursunny** [comment]
> 2\. It also relies on a service hosted by SwiftLaTeX to fetch these files instead of CTAN (even though I read somewhere that it _can_ use CTAN ...)

Neither would satisfy *cloud optional*.

If the file retrieval functionality is a pluggable module, it can be converted to retrieve from NDN network.
The entire CPAN library is then converted to a synchronized NDN repository, hosted on capable hosts.

**Comment by pulsejet** [comment]
> Neither would satisfy _cloud optional_.

It kinda does, if you can host your own repo of modules. We'll always have some webserver dependence anyway, unless we package this app with electron.

> If the file retrieval functionality is a pluggable module, it can be converted to retrieve from NDN network. The entire CPAN library is then converted to a synchronized NDN repository, hosted on capable hosts.

This is ideal 💯

**Comment by pulsejet** [comment]
I've marked the PR as ready @zjkmxy

File: src/components/share-latex/share-latex.tsx [hunk 1]
```diff
@@ -165,19 +166,21 @@ export default function ShareLatex(props: {
     window.open(fileUrl)
   }
 
-  let engine: any;
+  const [texEngine, setTexEngine] = createSignal<PdfTeXEngine>()
   const onCompile = async () => {
+    let engine = texEngine()
     if (!engine) {
-      engine = new (globalThis as any).PdfTeXEngine()
+      engine = new (globalThis as unknown as { PdfTeXEngine: { new(): PdfTeXEngine } }).PdfTeXEngine()
```
Comment (pulsejet) [inline src/components/share-latex/share-latex.tsx:173]: Ideally the clean way to do this is to augment the `global` module with the class type (say with `globals.d.ts`). But I couldn't get this to work with solidjs on first attempt.

**Comment by zjkmxy** [comment]
Based on our discussion we will not merge this PR for now, but will keep maintaining the branch until we make it a loadable module.

**Pull request merged**